src/dbnode/storage/shard.go

```diff
@@ -2616,74 +2615,149 @@ func (s *dbShard) Repair(
+	tileOpts := tile.Options{
+		FrameSize:          xtime.UnixNano(opts.Step.Nanoseconds()),
+		Start:              xtime.ToUnixNano(opts.Start),
+		UseArrow:           false,
+		ReaderIteratorPool: s.opts.ReaderIteratorPool(),
 	}
-	if err := reader.Open(openOpts); err != nil {
+
+	// TODO: these should probably be pooled
+	readerIter, err := tile.NewSeriesBlockIterator(crossBlockReader, tileOpts)
+	if err != nil {
 		return 0, err
 	}
-	defer reader.Close()
-
-	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
-	bytesReader := bytes.NewReader(nil)
-	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
-	var lastWriteError error
-	var processedBlockCount int64
 
-	for {
-		id, tags, data, _, err := reader.Read()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return processedBlockCount, err
+	closed := false
+	defer func() {
+		if !closed {
+			if err := readerIter.Close(); err != nil {
+				// NB: log the error on ungraceful exit.
+				s.logger.Error("could not close read iterator on error", zap.Error(err))
+			}
 		}
+	}()
 
-		data.IncRef()
-		bytesReader.Reset(data.Bytes())
-		dataPointIter.Reset(bytesReader, nil)
+	var (
+		processedBlockCount atomic.Int64
+		multiErr            xerrors.MultiError
+		downsampledValues   = make([]tile.DownsampledValue, 0, 4)
+	)
 
-		for dataPointIter.Next() {
-			dp, unit, annot := dataPointIter.Current()
-			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
-			if err != nil {
-				s.metrics.largeTilesWriteErrors.Inc(1)
-				lastWriteError = err
+	for readerIter.Next() {
+		seriesIter := readerIter.Current()
+		prevFrameLastValue := math.NaN()
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			id := frame.ID()
+			tags := frame.Tags()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+
+				downsampledValues = downsampledValues[:0]
+				lastIdx := len(frameValues) - 1
+
+				if opts.HandleCounterResets {
+					// last value plus possible few more datapoints to preserve counter semantics
+					tile.DownsampleCounterResets(prevFrameLastValue, frameValues, &downsampledValues)
+				} else {
+					// plain last value
+					downsampledValue := tile.DownsampledValue{FrameIndex: lastIdx, Value: frameValues[lastIdx]}
+					downsampledValues = append(downsampledValues, downsampledValue)
+				}
+
+				for _, downsampledValue := range downsampledValues {
+
+					value := downsampledValue.Value
+					timestamp := frame.Timestamps()[downsampledValue.FrameIndex]
+
+					if !singleUnit {
+						// TODO: what happens if unit has changed mid-tile?
+						// Write early and then do the remaining values separately?
+						unit = frame.Units().Values()[downsampledValue.FrameIndex]
```

Comment:
This will panic.

Reply:
Tagging @arnikola (not sure if you were subscribed to this PR notifications; also, not sure what is the original branch for this piece of code). I believe @gediminasgu has fixed it in his downstream branch, but still IMHO we should do a fix at the source so that we don't run into this again.